File: simplified-documents/src/main/java/org/nypl/simplified/documents/store/DocumentStoreType.java
package org.nypl.simplified.documents.store;

import com.io7m.jfunctional.OptionType;

import org.nypl.simplified.documents.authentication.AuthenticationDocumentType;
import org.nypl.simplified.documents.eula.EULAType;
import org.nypl.simplified.documents.synced.SyncedDocumentType;

/**
 * The type of document stores.
 */

public interface DocumentStoreType
{
  /**
   * @return The application privacy policy, if any.
   */

  OptionType<SyncedDocumentType> getPrivacyPolicy();

  /**
   * @return The application acknowledgements, if any.
   */

  OptionType<SyncedDocumentType> getAbout();

  /**
   * @return The application acknowledgements, if any.
   */

  OptionType<SyncedDocumentType> getAcknowledgements();

  /**
   * @return The login document
   */

  AuthenticationDocumentType getAuthenticationDocument();

  /**
   * @return The EULA, if any
   */

  OptionType<EULAType> getEULA();

  /**
   * @return The application licenses, if any.
   */

  OptionType<SyncedDocumentType> getLicenses();

}

File: simplified-documents/src/main/java/org/librarysimplified/documents/DocumentStoreType.kt
package org.librarysimplified.documents

import com.google.common.util.concurrent.ListenableFuture
import com.google.common.util.concurrent.ListeningExecutorService

interface DocumentStoreType {

  /**
   * @return The application privacy policy, if any.
   */

  val privacyPolicy: DocumentType?

  /**
   * @return The application acknowledgements, if any.
   */

  val about: DocumentType?

  /**
   * @return The application acknowledgements, if any.
   */

  val acknowledgements: DocumentType?

  /**
   * @return The EULA, if any
   */

  val eula: EULAType?

  /**
   * @return The application licenses, if any.
   */

  val licenses: DocumentType?

  /**
   * Run updates for all of the documents.
   */

  fun update(executor: ListeningExecutorService): ListenableFuture<*>
}
